src/components/evaluate.ts

```diff
@@ -0,0 +1,33 @@
+const expr = require('expression-eval');
+const cache = new Map();
+
+// Compile expression or return cached compiled expression
+function compile(str: string) {
+  const cached = cache.get(str);
```

Comment:
I don't like the fact we rely on strings here... references will be kept forever (I'd suppose so) here, so GC is unlikely to ever kick in here unless you explicitly clear the map or remove an item.
We could use memoize-one, as Vincenzo mentioned, though I suppose it may not make much sense given we kind of cache user's input here, so the cache would be invalidated every time user types sth.
The other solution would be to clear the map periodically... or use WeakMap, but I can't seem to find any object/fn we could use as a key here.

Reply:
> so the cache would be invalidated every time user types sth.

Maybe we could debounce the user input?

Reply:
Good idea!
We could also wait until input loses focus.

Reply:
I assumed so... because of that entire sandboxing mayhem.
If user is not allowed to provide his own code/expressions, what's the point of having a sandbox? :thinking: 
If one can choose from predefined values, we still don't need any advanced sandbox other than an isolated environment which wouldn't pollute global context (or local if eval is used)

Reply:
:-) Tehehe no, I just used `expression-eval` because it basically did exactly what I wanted (evaluate a string with substituting variables). That it happened to work using an interpreter was incidental really. I just chose it because a) it existed b) it worked well c) it was [2.4kb](https://bundlephobia.com/result?p=expression-eval@1.4.0). And most importantly, I'd forgotten JS _had_ a `with` statement.

Should we drop `expression-eval` for your cool code snippet?

Reply:
Not at all! I like `expression-eval`. It makes code look by far cleaner.
We could use my snippet only if it was placed in some different repo/package.
I mean, imagine, someone wants to make a change there and encounter such a quirk: with + Symbol.unscopables + Proxy. What would his/her reaction be?

Reply:
> What would his/her reaction be?

😱